kyaml/yaml/rnode.go

```diff
@@ -248,6 +443,391 @@ func (rn *RNode) IsNilOrEmpty() bool {
 		IsYNodeZero(rn.YNode())
 }
 
+// AsOrDie converts a RNode to the desired typed object. ptr must
+// be a pointer to a typed object. It will panic if it encounters an error.
+func (rn *RNode) AsOrDie(ptr interface{}) {
+	if err := rn.As(ptr); err != nil {
+		panic(err)
+	}
+}
+
+// As converts a RNode to the desired typed object. ptr must be
+// a pointer to a typed object.
+func (rn *RNode) As(ptr interface{}) error {
+	if rn.IsNil() {
+		return ErrRNodeNotFound{}
+	}
+	if ptr == nil || reflect.ValueOf(ptr).Kind() != reflect.Ptr {
+		return fmt.Errorf("ptr must be a pointer to an object")
+	}
+
+	// The built-in types only have json tags. We can't simply do mv.Node().Decode(ptr),
+	// since it use the lowercased field name by default if no yaml tag is specified.
+	// This affects both k8s built-in types (e.g. appsv1.Deployment) and any types
+	// that depends on built-in types (e.g. metav1.ObjectMeta, corev1.PodTemplate).
+	// To work around it, we rely on the json tags. We first convert mv to json
+	// and then unmarshal it to ptr.
+	j, err := rn.MarshalJSON()
+	if err != nil {
+		return err
+	}
+	err = json.Unmarshal(j, ptr)
+	return err
+}
+
+// Get gets the value for a nested field located by fields. A pointer must be
+// passed in, and the value will be stored in ptr. ptr can be any concrete type
+// (e.g. string, []corev1.Container, []string, corev1.Pod, map[string]string) or
+// a RNode. RNode should be used if you are dealing with comments that is more
+// than what LineComment, HeadComment, SetLineComment and SetHeadComment can
+// handle. It returns if the field is found and a potential error.
+func (rn *RNode) Get(ptr interface{}, fields ...string) (bool, error) {
```

Comment:
I have an idea for avoiding all the type checking in this one too. I'll share the code I was experimenting with at the bottom, but it isn't polished and I'm less sure of it than my other suggestions.

The thought driving the idea is that we only have the "can't use val.value.Decode" problem your comments repeatedly mention because of struct types that lack yaml tags. But most of the cases that we're handling _very clearly_ are not k8s core types, because they're primitive types! So can we use that fact to minimize the code paths we take / the manual work we need to do? _Somehow_, not necessarily the way I do it below.

The code below also has the (questionable?) benefit of being more flexible: if you give it a type that *can* be expressed in the ptr type given, it will automatically be coerced into it. E.g. the test where you try to retrieve spec.replicas as a string just works.

LMK what you think. 

```go
func (rn *RNode) Get(ptr interface{}, fields ...string) (bool, error) {
	found, err := func() (bool, error) {
		if rn.IsNil() {
			return false, ErrRNodeNotFound{}
		}

		if ptr == nil || reflect.ValueOf(ptr).Kind() != reflect.Ptr {
			return false, fmt.Errorf("ptr must be a pointer to an object")
		}

		foundVal, found, err := rn.nestedValue(fields...)
		if err != nil || !found {
			return found, err
		}
		inputAsDocRNode, err := NewRNodeFrom(ptr)
		if err != nil {
			return false, err
		}
		inputAsRNode := NewRNode(inputAsDocRNode.value.Content[0])
		// trick to try to detect fundamental incompatibilities
		if foundVal.value.Kind != inputAsRNode.value.Kind {
			return true, errors.Errorf("needed a %v to get field %s into ptr, got %v", foundVal.value.Kind, fields, inputAsRNode.value.Kind) // this message is terrible, need to translate Kind ints to strings
		}

		switch foundVal.value.Kind {
		case yaml.MappingNode, yaml.SequenceNode:
			if err := foundVal.As(ptr); err != nil {
				return true, err
			}
			return true, nil
		case yaml.ScalarNode:
			return true, foundVal.value.Decode(ptr)
		default:
			return true, errors.Errorf("unhandled kind %v", foundVal.value.Kind)
		}
	}()
	if err != nil {
		return found, fmt.Errorf("unable to get fields %v as %T with error: %w", fields, ptr, err)
	}
	return found, nil
}
```